demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+ */
+/* #define REPROVISION_EACH_TIME 0 */
+
+
+#define pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS    "Device Priv TLS Key"
+#define pkcs11testLABEL_DEVICE_PRIVATE_KEY_FOR_TLS      pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS
+
+#define pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS     "Device Pub TLS Key"
+#define pkcs11testLABEL_DEVICE_PUBLIC_KEY_FOR_TLS       pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS
+
+#define SHA256_DIGEST_SIZE                              32
+#define ECDSA_SIGNATURE_SIZE                            64
+#define RSA_SIGNATURE_SIZE                              256
+#define MBEDTLS_X509_CSR_H
+#define MBEDTLS_X509_CSR_WRITE_C
+#define MBEDTLS_PEM_WRITE_C
+
+typedef enum
+{
+    eNone,                /* Device is not provisioned.  All credentials have been destroyed. */
+    eRsaTest,             /* Provisioned using the RSA test credentials located in this file. */
+    eEllipticCurveTest,   /* Provisioned using EC test credentials located in this file. */
+    eClientCredential,    /* Provisioned using the credentials in aws_clientcredential_keys. */
+    eGeneratedEc,         /* Provisioned using elliptic curve generated on device.  Private key unknown.  No corresponding certificate. */
+    eGeneratedRsa,
+    eDeliberatelyInvalid, /* Provisioned using credentials that are meant to trigger an error condition. */
+    eStateUnknown         /* State of the credentials is unknown. */
+} CredentialsProvisioned_t;
+
+
+
+/* GLOBAL VARIABLES */
+CK_SESSION_HANDLE xGlobalSession;
+CK_RV xResult;
+
+CK_OBJECT_HANDLE xPrivateKeyHandle = CK_INVALID_HANDLE;
+CK_OBJECT_HANDLE xPublicKeyHandle = CK_INVALID_HANDLE;
+CK_FUNCTION_LIST_PTR pxGlobalFunctionList;
+
+/*-----------------------------------------------------------*/
+
+/* Declaration of demo function. */
+void vStartKeyProvisioningDemo( void );
+
+/*-----------------------------------------------------------*/
+
+
+/* @brief Random Number Generator Used to Generate CSR
+ *
+ *
+ *	\param[in] pkcs_session     Pointer to PKCS Session Handle
+ *
+ *  \param[in] pucRandom        Starting Point
+ *
+ *  \param[in] xRandomLength    Length
+ *
+ *  \return 0 on success
+ *
+ */
+static int prvRNG( void * pkcs_session,
+                   unsigned char * pucRandom,
+                   size_t xRandomLength )
+{
+    BaseType_t xResult;
+    CK_FUNCTION_LIST_PTR pxP11FunctionList;
+
+    xResult = C_GetFunctionList( &pxP11FunctionList );
+
+    xResult = pxP11FunctionList->C_GenerateRandom( ( *( ( CK_SESSION_HANDLE * ) pkcs_session ) ), pucRandom, xRandomLength );
+
+    if( xResult != 0 )
+    {
+        configPRINTF( ( "ERROR: Failed to generate random bytes %d \r\n", xResult ) );
+        xResult = TLS_ERROR_RNG;
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+/* @brief Alternate Signing Function to be Passed into PK Context Header */
+static int prvPrivateKeySigningCallback( void * pvContext,
```

Comment:
Is this copied from somewhere else? That's understandable from the perspective of expedience for this demo. But it would be helpful to have a comment about where it came from, in case there's subsequent interest in refactoring.

Reply:
Yes, it comes from iot_tls.c. Just added a comment to clarify.